Add unicode handling capability

vleni opened the pull request:
Previously, if the client program outputs unicode, the unicode output would be garbled when it is piped through `NBReader`. For more detail, see #105. Writing doesn't seem to have any trouble as demonstrated by `examples/cat.rs`.

Right now the unicode encoder is always on. In pexpect, an encoding option is available to toggle between the two, which I'm not sure where to put here.

vleni commented:
The current behaviour is to withhold a half-completed unicode char from the output buffer. If the client program outputs EOF when the unicode char buffer is incomplete, the char is swallowed. However one problem is that if the client program outputs an invalid unicode char and then valid unicode chars, the output buffer will be stalled and no more chars will be piped into the buffer.

matthiasbeyer commented:
Thanks for your patches and also for your bug report! :heart: Highly appreciated!

The CI failure should be fixed after #107 is merged, don't worry about that.

I am not sure what the way forward is yet. I like the first option you proposed in #105 most:

> Change the type of PipedChar(u8) to PipedChar(char): If the program sends over half of a unicode char and then stop it would hang the reader

But that might break some downstream users for non-UTF8 outputting programs, as I understand it!? @petreeftime maybe you have some ideas what's the best way to go here? Maybe make the whole thing configurable (a flag that can be passed to the library for each call)?

vleni commented:
> Thanks for your patches and also for your bug report! ❤️ Highly appreciated!
> 
> The CI failure should be fixed after #107 is merged, don't worry about that.
> 
> I am not sure what the way forward is yet. I like the first option you proposed in #105 most:
> 
> > Change the type of PipedChar(u8) to PipedChar(char): If the program sends over half of a unicode char and then stop it would hang the reader
> 
> But that might break some downstream users for non-UTF8 outputting programs, as I understand it!? @petreeftime maybe you have some ideas what's the best way to go here? Maybe make the whole thing configurable (a flag that can be passed to the library for each call)?

I don't think the first option is the best one. The solution in this PR is kind of a compromise between the first two options. This is because the underlying datatype piped by the client program is always `u8` and not some `char` type, and any assembly of `u8` into `char`s is only really needed when downstream needs to check the existence of certain `char`s to e.g. break line or search a regex.

You're correct that this may break non-UTF8 outputting programs, but if the program restricts its output to the ASCII range there shouldn't be any compatibility issue. I left a flag in the code to represent some external option which toggles between unicode and non-unicode encoding.

Do you think it would be sensible to add it as a crate-level feature? Or embed the option in every instance of `NBReader` and hence `spawn` (this is the solution from pexpect)?

matthiasbeyer commented:
If anything I'd have it as an option on the data type, not a crate-level feature or compiletime option, because of the simple fact that someone might have UTF-8 outputting programs and non-UTF-8 outputting programs in one project. In general I think compiletime features should never be used for "either-or" functionality :laughing: 

So a flag on the type would be totally fine with me, with UTF-8 compatible reading as the default.

I'd like to see what @petreeftime thinks as well though.

vleni commented:
> If anything I'd have it as an option on the data type, not a crate-level feature or compiletime option, because of the simple fact that someone might have UTF-8 outputting programs and non-UTF-8 outputting programs in one project. In general I think compiletime features should never be used for "either-or" functionality 😆
> 
> So a flag on the type would be totally fine with me, with UTF-8 compatible reading as the default.
> 
> I'd like to see what @petreeftime thinks as well though.

If you have any idea for where to add such a flag I can do it in this PR.

By the way why is [Pull Request Checks](https://github.com/rust-cli/rexpect/actions/runs/5887321624/job/15967353122?pr=106) failing? Is it because I didn't sign my commits?

matthiasbeyer commented:
> By the way why is [Pull Request Checks](https://github.com/rust-cli/rexpect/actions/runs/5887321624/job/15967353122?pr=106) failing? Is it because I didn't sign my commits?

Its because you didn't [signoff](https://stackoverflow.com/questions/1962094/what-is-the-sign-off-feature-in-git-for) your commits (the `-s` flag in `git-commit`). You can fix this by doing something like `git rebase $(git merge-base origin/master HEAD) -x "git commit --amend --no-edit -s"` (there might be a more elegant way... :shrug:)

matthiasbeyer commented:
We merged #107 that should make your compile issues go away, please do rebase to latest master.

vleni commented:
> We merged #107 that should make your compile issues go away, please do rebase to latest master.

Rebased. I added a field for `NBReader` which contains the encoding. The encoding cannot be set via a setter because by the time the setter is called the receiving thread may have already received one or two chars, so the encoding has to be known by the time `NBReader::new` is called. Whether you would like this function argument to `NBReader::new` to be rippled to everything else that uses `NBReader::new` is your call.

matthiasbeyer commented on src/encoding.rs:
```diff
@@ -0,0 +1,6 @@
+#[derive(Debug)]
```
We should derive `Eq` and `PartialEq` here as well. It doesn't hurt and because the type is exposed our users might need these.

vleni replied:
fixed

matthiasbeyer reviewed (changes requested):
Overall this looks good.

I think having the encoding be an argument on `new`, but also providing two convenience helpers `NBReader::ascii()` and `NBReader::utf8()`, which call `new` with the respective argument would be nice and not that much of a hassle.

@petreeftime pinging you again, hope you can have a look as well.

vleni commented:
> Overall this looks good.
> 
> I think having the encoding be an argument on `new`, but also providing two convenience helpers `NBReader::ascii()` and `NBReader::utf8()`, which call `new` with the respective argument would be nice and not that much of a hassle.
> 
> @petreeftime pinging you again, hope you can have a look as well.

Have you decided that this is the way to go? I need this crate in production

vleni commented:
I've merged `master` into this branch and propagated `encoding` into the new `Options` structure. By default it still uses `ASCII` encoding everywhere. However I'm getting this error:
```
---- session::tests::test_bash stdout ----
thread 'session::tests::test_bash' panicked at 'assertion failed: `(left == right)`
  left: `"/tmp\r\n"`,
 right: `"\u{1b}[?2004l\r\r\n/tmp\r\n\u{1b}[?2004h"`', src/session.rs:543:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```
it seems like bash is generating some extra outputs? Is this related to my changes?

ccntrq commented:
Hi all,

I’d love to see Unicode (UTF-8) support land in rexpect! To help move this PR forward, I’ve merged the latest changes from main into my own copy of this branch ([ccntrq/rexpect-unicode/tree/unicode](https://github.com/ccntrq/rexpect-unicode/tree/unicode)) (Didn't rebase to preserve the signed commits by @lenianiva). The [test failures](https://github.com/rust-cli/rexpect/pull/106#issuecomment-1972528919) previously mentioned by @lenianiva are now resolved in `main`.

While updating, I also addressed a bug in `try_read` related to reading into the middle of a multi-byte UTF-8 character (see: https://github.com/ccntrq/rexpect-unicode/commit/25988477f5815922469fddbc91e217f4960aae68).

Notable changes in my updated branch:
- ASCII remains the default encoding for now, to retain current behavior.
- UTF-8 mode is available via `spawn_with_options` At the moment, that’s the only way to opt into Unicode support.
- I’d suggest we add convenience wrappers (e.g., `spawn_utf8`, `spawn_command_utf8`, etc.) so users can easily spawn sessions in UTF-8 mode without having to manually specify options.

What do you think about this API direction? Are there any further changes or cleanup needed before merging? I’m happy to help finish this up!

Thanks for all your work on rexpect.
